Bluetooth: Mesh: Add FWID to DFU metadata json (#20255)
nrfconnect/sdk-nrf: doc/nrf/protocols/bt/bt_mesh/dfu_over_bt_mesh.rst

By enabling the ``SB_CONFIG_DFU_ZIP_BLUETOOTH_MESH_METADATA`` option in sysbuild, the metadata will be automatically parsed from the ``.elf`` and ``.config`` files.
The parsed data is stored in the :file:`ble_mesh_metadata.json` file.
The file is placed in the :file:`dfu_application.zip` archive in the build folder of the application.

> Additionally, the metadata string and firmware ID required by the ``mesh models dfu slot add`` command will be printed in the command line window when the application is built::

Reviewer: ```suggestion
Additionally, the metadata string and firmware ID required by the ``mesh models dfu slot add`` command is printed in the command-line window when the application is built::
```

  Bluetooth Mesh Composition metadata generated:
    Encoded metadata: 020000000000000094cf24017c26f3710100
    Firmware ID: 59000200000000000000
    Full metadata written to: APPLICATION_FOLDER\build\zephyr\dfu_application.zip

The Firmware ID can be generated from one of multiple sources:

>   * To use a user-supplied hex-string as the Firmware ID, enable the ``SB_CONFIG_DFU_ZIP_BLUETOOTH_MESH_METADATA_FWID_CUSTOM`` option. This is the default behavior.

Author: @peknis For some reason this line ends up in boldface. I don't think I understand why, do you have any idea? (See doc build output, https://github.com/nrfconnect/sdk-nrf/pull/20255)

Reviewer: Yep, my bad. Start the first bullet from the very beginning of the line and the sub bullet as usual from the beginning of the previous bullet's text.

>       * If this option is selected, the ``SB_CONFIG_DFU_ZIP_BLUETOOTH_MESH_METADATA_FWID_CUSTOM_HEX`` option must be set to a valid Firmware ID.

Author: ````When using the :ref:`zephyr:bluetooth_mesh_dfd_srv` for distribution, the number of bytes in the Firmware ID of images used during distribution must not exceed the value of the ``CONFIG_BT_MESH_DFU_FWID_MAXLEN`` option set on the distributor node.````?

>   * To use a user-supplied hex-string as the Firmware ID, enable the ``SB_CONFIG_DFU_ZIP_BLUETOOTH_MESH_METADATA_FWID_CUSTOM`` option. This is the default behavior.
>       * If this option is selected, the ``SB_CONFIG_DFU_ZIP_BLUETOOTH_MESH_METADATA_FWID_CUSTOM_HEX`` option must be set to a valid Firmware ID.
>         At minimum it should include a Company ID, the rest of the string is vendor-specific version information.

Reviewer: ```suggestion
  * To use a user-supplied hex-string as the Firmware ID, enable the ``SB_CONFIG_DFU_ZIP_BLUETOOTH_MESH_METADATA_FWID_CUSTOM`` option.
    This is the default behavior.
      * If this option is selected, the ``SB_CONFIG_DFU_ZIP_BLUETOOTH_MESH_METADATA_FWID_CUSTOM_HEX`` option must be set to a valid Firmware ID.
        At minimum, it should include a Company ID.
        The rest of the string is vendor-specific version information.
```

>   * To use a Firmware ID consisting of the value of ``CONFIG_BT_COMPANY_ID`` and the MCUboot version, enable the ``SB_CONFIG_DFU_ZIP_BLUETOOTH_MESH_METADATA_FWID_MCUBOOT_VERSION`` option.

Reviewer: ```suggestion
* To use a Firmware ID consisting of the value of ``CONFIG_BT_COMPANY_ID`` and the MCUboot version, enable the ``SB_CONFIG_DFU_ZIP_BLUETOOTH_MESH_METADATA_FWID_MCUBOOT_VERSION`` option.
```

.. note::

>    It is required that the Firmware ID inserted into the JSON file matches the ``fwid`` used when initializing the :ref:`zephyr:bluetooth_mesh_dfu_srv`.
>    Application developers are responsible for ensuring this.

Reviewer: ```suggestion
   The Firmware ID inserted into the JSON file must match the ``fwid`` used when initializing the :ref:`zephyr:bluetooth_mesh_dfu_srv`.
```

By enabling the ``SB_CONFIG_DFU_ZIP_BLUETOOTH_MESH_METADATA`` option in sysbuild, the metadata will be automatically parsed from the ``.elf`` and ``.config`` files.
The parsed data is stored in the :file:`ble_mesh_metadata.json` file.
The file is placed in the :file:`dfu_application.zip` archive in the build folder of the application.
Additionally, the metadata string and firmware ID required by the ``mesh models dfu slot add`` command is printed in the command-line window when the application is built::

  Bluetooth Mesh Composition metadata generated:
    Encoded metadata: 020000000000000094cf24017c26f3710100
    Firmware ID: 59000200000000000000
    Full metadata written to: APPLICATION_FOLDER\build\zephyr\dfu_application.zip

> The Firmware ID can be generated from one of multiple sources:

Reviewer 2: Something seems strange to read here.

```suggestion
The Firmware ID can be generated from one of the multiple sources:
```
or

```suggestion
The Firmware ID can be generated using one of the options mentioned below:
```

Author: "The Firmware ID can be generated from one of _the_ multiple sources" is objectively worse grammar, isn't it? @peknis ?

Reviewer: Yes, the article was not meant to be there. Omkar's suggestion sounds better.

> * To use a user-supplied hex-string as the Firmware ID, enable the ``SB_CONFIG_DFU_ZIP_BLUETOOTH_MESH_METADATA_FWID_CUSTOM`` option.
>   This is the default behavior.
>   * If this option is selected, the ``SB_CONFIG_DFU_ZIP_BLUETOOTH_MESH_METADATA_FWID_CUSTOM_HEX`` option must be set to a valid Firmware ID.
>     At minimum, it should include a Company ID in little-endian order.
>     The rest of the string is vendor-specific version information.
>   * When using the :ref:`zephyr:bluetooth_mesh_dfd_srv` for distribution, the number of bytes in the Firmware ID of images used during distribution must not exceed the value of the ``CONFIG_BT_MESH_DFU_FWID_MAXLEN`` option set on the distributor node.
> * To use a Firmware ID consisting of the value of the ``CONFIG_BT_COMPANY_ID`` option and the value of the ``CONFIG_MCUBOOT_IMGTOOL_SIGN_VERSION`` option, enable the ``SB_CONFIG_DFU_ZIP_BLUETOOTH_MESH_METADATA_FWID_MCUBOOT_VERSION`` option.

Reviewer: ```suggestion
* To use a user-supplied hex-string as the Firmware ID, enable the ``SB_CONFIG_DFU_ZIP_BLUETOOTH_MESH_METADATA_FWID_CUSTOM`` option.
  This is the default behavior.

  * If this option is selected, the ``SB_CONFIG_DFU_ZIP_BLUETOOTH_MESH_METADATA_FWID_CUSTOM_HEX`` option must be set to a valid Firmware ID.
    At minimum, it should include a Company ID in little-endian order.
    The rest of the string is vendor-specific version information.
  * When using the :ref:`zephyr:bluetooth_mesh_dfd_srv` for distribution, the number of bytes in the Firmware ID of images used during distribution must not exceed the value of the ``CONFIG_BT_MESH_DFU_FWID_MAXLEN`` option set on the distributor node.

* To use a Firmware ID consisting of the value of the ``CONFIG_BT_COMPANY_ID`` option and the value of the ``CONFIG_MCUBOOT_IMGTOOL_SIGN_VERSION`` option, enable the ``SB_CONFIG_DFU_ZIP_BLUETOOTH_MESH_METADATA_FWID_MCUBOOT_VERSION`` option.
```

.. note::
   The Firmware ID inserted into the JSON file must match the ``fwid`` used when initializing the :ref:`zephyr:bluetooth_mesh_dfu_srv`.

> 

Reviewer 2: This note confuses me. When does the Firmware ID inserted into the JSON file must match?
1. At the time of first boot?
2. At the time of processing of Firmware Update Start? or in-between (like metadata checks, etc.)
3. At the time of attempting a confirm/verify stage or SWAP?

Can you please clarify this note?

Author: It's a bit difficult. The point is that, whenever you run "west build", any code that calls `BT_MESH_DFU_SRV_INIT(..., struct_with_pointer_to_fwid, ...)` and the value set in Kconfig for the `SB_CONFIG_DFU_ZIP_BLUETOOTH_MESH_METADATA_FWID_...` options should use the same FWID in those two locations. I don't know how to clarify this properly.

Reviewer 3: don't know how this system works but why can't the sysbuild value be propagated to applications which then directly use it instead of having a potential mismatch?

Author: > don't know how this system works but why can't the sysbuild value be propagated to applications which then directly use it instead of having a potential mismatch?

Because we are making a minimal solution here.

Reviewer 3: Can the value in the application be replaced with a Kconfig? If so I can knock up a PR to propagate the value from sysbuild to application in a couple of minutes

Author: Bit longer explanation: The API exposed to the applications currently is very open-ended and lets applications supply whatever byte string they want as FWID in the image list. To enforce a match between application behavior and JSON contents would mean several breaking API changes, which we need to think about before implementing.

Author: > Can the value in the application be replaced with a Kconfig? If so I can knock up a PR to propagate the value from sysbuild to application in a couple of minutes

Yes sure, but we can't force every existing or future application to use the Kconfig in their code. The API they have access to lets them do whatever they want.

Author: @omkar3141 I believe this is the only discussion left. Do you have more input after my explanation(s) above? I struggle to formulate this in an efficient and clear way.

Reviewer 2: How about this? 
```suggestion
.. note::
   The Firmware ID (FWID) specified in the JSON file must match the ``fwid`` used when initializing the :ref:`zephyr:bluetooth_mesh_dfu_srv`. This ensures that after the firmware update is applied, the node will respond with the correct FWID in the Firmware Update Information Status message.

```

Reviewer: And again, remember to have each sentence on a separate line.

By enabling the ``SB_CONFIG_DFU_ZIP_BLUETOOTH_MESH_METADATA`` option in sysbuild, the metadata will be automatically parsed from the ``.elf`` and ``.config`` files.
The parsed data is stored in the :file:`ble_mesh_metadata.json` file.
The file is placed in the :file:`dfu_application.zip` archive in the build folder of the application.

> Additionally, the metadata string and firmware ID required by the ``mesh models dfu slot add`` command is printed in the command-line window when the application is built::

Reviewer: ```suggestion
Additionally, the metadata string and firmware ID required by the ``mesh models dfu slot add`` command are printed in the command-line window when the application is built::
```

  Bluetooth Mesh Composition metadata generated:
    Encoded metadata: 020000000000000094cf24017c26f3710100

>     Firmware ID: 59000200000000000000

Reviewer 2: May be worth doing this if you can spare couple of extra minutes. It is rather worth showing example with following numbers to show endianness conversion with action.

    fwid.company_id = 0x0059;
    fwid.major = 0xAA;
    fwid.minor = 0xBB;
    fwid.revision = 0x1234;
    fwid.build_num = 0x89ABCDEF;
    
  Which will become "5900aabb3412efcdab89"

Author: Users shouldn't need to care about the endianness of the FWID_MCUBOOT_VERSION IMO. Besides, the "5900" company ID already shows the little-endianness.

Reviewer 4: > Users shouldn't need to care about the endianness of the FWID_MCUBOOT_VERSION IMO

I'm still uncertain about this. The model spec (I reffered initially to Access layer, but model layer has the same requirement) requires all data to be encoded into little endianness (section 3.8.1). The DFU server in our stack simply picks what points to fwid and passes it over the mesh network:
https://github.com/zephyrproject-rtos/zephyr/blob/baced6d315f0c58cf4950a8c49cb7a49d18c208f/subsys/bluetooth/mesh/dfu_srv.c#L171-L172

This works fine with our stack because we pass the fwid "as is" (as it comes in the message). But if some other stack is used, it can convert fwid into system endianness, and we will have problems here.

Author: @PavelVPV this was in response to Omkars suggestion about demonstrating endianness (of the MCUBOOT FWID!) in the user guide sample output. This is already little endian in my current implementation. I think you're confusing it with our separate discussion on endianness for the CUSTOM fwid.

    Full metadata written to: APPLICATION_FOLDER\build\zephyr\dfu_application.zip

The Firmware ID can be generated from one of multiple sources:

* To use a user-supplied hex-string as the Firmware ID, enable the ``SB_CONFIG_DFU_ZIP_BLUETOOTH_MESH_METADATA_FWID_CUSTOM`` option.
  This is the default behavior.

  * If this option is selected, the ``SB_CONFIG_DFU_ZIP_BLUETOOTH_MESH_METADATA_FWID_CUSTOM_HEX`` option must be set to a valid Firmware ID.
    At minimum, it should include a Company ID in little-endian order.
    The rest of the string is vendor-specific version information.

>   * When using the :ref:`zephyr:bluetooth_mesh_dfd_srv` for distribution, the number of bytes in the Firmware ID of images used during distribution must not exceed the value of the ``CONFIG_BT_MESH_DFU_FWID_MAXLEN`` option set on the distributor node.

Reviewer: ```suggestion
  * When using the :ref:`zephyr:bluetooth_mesh_dfd_srv` for distribution, the number of bytes in the Firmware ID of images used during distribution must not exceed the value of the :kconfig:option`CONFIG_BT_MESH_DFU_FWID_MAXLEN` Kconfig option set on the distributor node.
```

> * To use a Firmware ID consisting of the value of the ``CONFIG_BT_COMPANY_ID`` option and the value of the ``CONFIG_MCUBOOT_IMGTOOL_SIGN_VERSION`` option, enable the ``SB_CONFIG_DFU_ZIP_BLUETOOTH_MESH_METADATA_FWID_MCUBOOT_VERSION`` option.

Reviewer: ```suggestion
* To use a Firmware ID consisting of the values of the :kconfig:option:`CONFIG_BT_COMPANY_ID` and of the :kconfig:option:`CONFIG_MCUBOOT_IMGTOOL_SIGN_VERSION` Kconfig options, enable the ``SB_CONFIG_DFU_ZIP_BLUETOOTH_MESH_METADATA_FWID_MCUBOOT_VERSION`` option.
```

> * To use a Firmware ID consisting of the values of the :kconfig:option:`CONFIG_BT_COMPANY_ID` option and the :kconfig:option:`CONFIG_MCUBOOT_IMGTOOL_SIGN_VERSION` Kconfig options, enable the ``SB_CONFIG_DFU_ZIP_BLUETOOTH_MESH_METADATA_FWID_MCUBOOT_VERSION`` option.

Reviewer: ```suggestion
* To use a Firmware ID consisting of the values of the :kconfig:option:`CONFIG_BT_COMPANY_ID`  and the :kconfig:option:`CONFIG_MCUBOOT_IMGTOOL_SIGN_VERSION` Kconfig options, enable the ``SB_CONFIG_DFU_ZIP_BLUETOOTH_MESH_METADATA_FWID_MCUBOOT_VERSION`` option.
```

By enabling the ``SB_CONFIG_DFU_ZIP_BLUETOOTH_MESH_METADATA`` option in sysbuild, the metadata will be automatically parsed from the ``.elf`` and ``.config`` files.
The parsed data is stored in the :file:`ble_mesh_metadata.json` file.
The file is placed in the :file:`dfu_application.zip` archive in the build folder of the application.

> Additionally, the metadata string and firmware ID required by the ``mesh models dfu slot add`` command are printed in the command-line window when the application is built::

Reviewer 4: > and firmware ID required by...

Only if either `SB_CONFIG_DFU_ZIP_BLUETOOTH_MESH_METADATA_FWID_CUSTOM` or `SB_CONFIG_DFU_ZIP_BLUETOOTH_MESH_METADATA_FWID_MCUBOOT_VERSION` is enabled, right (`DFU_ZIP_BLUETOOTH_MESH_METADATA_FWID` is optional)?

  Bluetooth Mesh Composition metadata generated:
    Encoded metadata: 020000000000000094cf24017c26f3710100
    Firmware ID: 59000200000000000000
    Full metadata written to: APPLICATION_FOLDER\build\zephyr\dfu_application.zip

You can generate the Firmware ID using one of the following options:

* To use a user-supplied hex-string as the Firmware ID, enable the ``SB_CONFIG_DFU_ZIP_BLUETOOTH_MESH_METADATA_FWID_CUSTOM`` option.
  This is the default behavior.

  * If this option is selected, the ``SB_CONFIG_DFU_ZIP_BLUETOOTH_MESH_METADATA_FWID_CUSTOM_HEX`` option must be set to a valid Firmware ID.
    At minimum, it should include a Company ID in little-endian order.
    The rest of the string is vendor-specific version information.

>   * When using the :ref:`zephyr:bluetooth_mesh_dfd_srv` for distribution, the number of bytes in the Firmware ID of images used during distribution must not exceed the value of the :kconfig:option:`CONFIG_BT_MESH_DFU_FWID_MAXLEN` Kconfig option set on the distributor node.

Reviewer 2: Potential task. I think we should not have `CONFIG_BT_MESH_DFU_FWID_MAXLEN` option to begin with, and ideally distributor should support max length as defined by the spec. Because distributor could be used in a mixed ecosystem, and then target won't be upgradable using a specific distributor and a specific vendor app, IF target uses longer FWID. That would be really annoying situation to have to just save 100 or so bytes of RAM.

Author: Feel free to create a JIRA task
